### PR TITLE
Add DE component

### PR DESCRIPTION
**Pull request by kelshmo**
This PR computes differential expression with the `variancePartition` packages and adds the respective dependencies. The function will be added to the `drake` plan in a subsequent pull request, once the intermediate steps are accounted for. 
## Making the model 
The input model is determined by the user-provided `primary_variable` and `model_variables`. The `primary_variable` is modeled as a fixed effect. In the case multiple variables are provided (e.g. `c("var1", "var2")` the variables are modeled with an interaction term, and still as a fixed effect. The metadata is updated accordingly with the `build_formula` function. 
## Contrasts
Contrasts are defined automatically - all unique combinations of all levels of the `primary_variable` are tested as a contrast using `variancePartition::dream`

## DE
All differential expression results for each contrast are merged into a single data frame.

## Output
- the `voom` object will be input for a violin plot visualizing variance per variable
- `contrasts` will be visualized
- adjusted fit for computing residuals
- differential expression results for each contrast

### CODE REVIEW

**Comment by kelshmo** [comment]
@karawoo I'm hoping to get your input specifically on the [test I wrote for this function](https://github.com/Sage-Bionetworks/sageseqr/pull/66/files#diff-229e7f464435919d7a238cc5dda9bbb0) - I need to mock a dataset of a certain size in order for the function to run. I'm not sure it is the best idea to depend on a test dataset from a package (`varianceParition` in this case). Is there a standard for dealing with this kind of use cases in tests (where data is imported)?

**Comment by karawoo** [comment]
Yeah, I would not recommend using external data from another package in the test as it's probably more complex than is ideal, and it could change and break your test even if all of your code still works correctly. Can you construct a minimal fake dataset within your test ([example](https://github.com/Sage-Bionetworks/dccvalidator/blob/35c61877181f155fe3c7692df60c42e3e890d3dd/tests/testthat/test-check-ages-over-90.R#L3-L13))?

**Comment by kelshmo** [comment]
Thanks @karawoo. I updated the test example to not rely on external data and it should work fine. Currently, the test is failing due to an issue in `variancePartition`:https://github.com/GabrielHoffman/variancePartition/issues/15

**Comment by kelshmo** [comment]
The bug has been fixed and I updated the DESCRIPTION to point to a [specific version](https://github.com/Sage-Bionetworks/sageseqr/pull/66/files#diff-35ba4a2677442e210c23a00a5601aba3R50) and [commit](https://github.com/Sage-Bionetworks/sageseqr/pull/66/files#diff-35ba4a2677442e210c23a00a5601aba3R56) of the package but I'm perplexed - the test continue to fail even when it passes with `devtools::test()` locally and runs fine in my local environment with `variancePartition 1.19.10`.  `sageseqr` must not be pointing to the correct version of `variancePartition`. @karawoo  do you have any debugging ideas?

File: tests/testthat/test-de.R [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                             continuous = c("RIN", "age"),
+                             sample_identifier = c("samples")
+)
+
+biomart_results <- tibble::column_to_rownames(
+  tibble::tibble(ensembl_gene_id = rownames(counts), length = 400),
+  var = "ensembl_gene_id"
+)
+de <- differential_expression(counts,
+                              counts,
```
Comment (karawoo) [inline tests/testthat/test-de.R:37]: Is `counts` supposed to be here twice?

Comment (karawoo) [reply tests/testthat/test-de.R:37]: Never mind, I see that there are two different arguments for counts

**Comment by karawoo** [comment]
The problem is with this line in `voomWithDreamWeights`: https://github.com/GabrielHoffman/variancePartition/blob/e61a3de6aa94c4a429f0147bdbdae4f7f7a915b7/R/voomWithDreamWeights.R#L65

I don't know what `design` is supposed to be, but it's not an argument to the function and is not defined anywhere before it is used.

**Comment by kelshmo** [comment]
@jgockley62 The bug with `variancePartition` is resolved. Do you have any further comments?

File: tests/testthat/test-de.R [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+test_that("output is a list of 6", {
+  expect_equal(length(de), 6)
+})
+
+test_that("voom output is complete", {
+  expect_equal(names(de$voom_object), c("targets", "E", "weights"))
+})
+
+test_that("formula is complete", {
+  expect_false(grepl("NULL", de$formula))
```
[the remark this answers is not in the document]
Comment (kelshmo) [reply tests/testthat/test-de.R:42]: I want to test whether the string contains `"NULL"`.

File: R/functions.R [hunk 3]
```diff
@@ -369,3 +369,137 @@ mclust::Mclust
...
+#' to scale numeric variables and model categorical variables as a random
+#' effect. Additionally, variables are scaled to account for
+#' multiple variables that might have an order of magnitude difference.
+#'
+#' @param model_variables Vector of variables to include in the linear (mixed) model.
+#' @param primary_variable Vector of variables that will be collapsed into a single
+#' fixed effect interaction term.
+#' @inheritParams coerce_factors
+#' @export
+build_formula <- function(md, model_variables, primary_variable) {
```
[the remark this answers is not in the document]
Comment (kelshmo) [reply R/functions.R:385]: @karawoo Do you have a suggestion for dealing with multiple classes? I implemented  `purrr::map` in place of `sapply`.

[the remark this answers is not in the document]
Comment (kelshmo) [reply R/functions.R:385]: Interesting with this example, only the first `class` is evaluated.

``` r
dat <- data.frame(x = 1, y = "2020-01-01")
class(dat$x) <- c("special numeric", "numeric")
sapply(dat, class) %in% c("factor", "numeric")
#> [1] FALSE FALSE
```

<sup>Created on 2020-07-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

[the remark this answers is not in the document]
Comment (kelshmo) [reply R/functions.R:385]: Maybe, checking for `any` TRUE is an approach? 
``` r
dat <- data.frame(x = 1, y = "2020-01-01")
class(dat$x) <- c("special numeric", "numeric")
tmp_classes <- purrr::map(dat, class) 
bar <- purrr::map(tmp_classes, ~ .x %in% c("factor", "numeric"))
purrr::map(bar, any)
#> $x
#> [1] TRUE
#> 
#> $y
#> [1] FALSE
```

<sup>Created on 2020-07-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

[the remark this answers is not in the document]
Comment (karawoo) [reply R/functions.R:385]: Another option would be to use `inherits()`: 
``` r
dat <- data.frame(x = 1, y = "2020-01-01")
class(dat$x) <- c("special numeric", "numeric")
purrr::map_lgl(dat, ~ inherits(.x, c("numeric", "factor")))
#>     x     y 
#>  TRUE FALSE
```

<sup>Created on 2020-07-24 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

[the remark this answers is not in the document]
Comment (kelshmo) [reply R/functions.R:385]: Nice, so concise! I added this function to `build_formula`. There will be issues with subsequent steps that I documented to fix in #72.

**Pull request merged**